libp2p/switch.nim

```diff
@@ -288,6 +289,12 @@ proc stop*(s: Switch) {.async, public.} =
 
   s.started = false
 
+  try:
+    # Stop accepting incoming connections
+    await allFutures(s.acceptFuts.mapIt(it.cancelAndWait())).wait(1.seconds)
+  except CatchableError as exc:
+    debug "Cannot cancel accepts", error = exc.msg
+
```

Comment:
It seems to be a bit related https://github.com/vacp2p/nim-libp2p/pull/1065#discussion_r1561355153

Reply:
yes - though relying on `transport.running` runs into synchronization problems. 

The way to think about futures and indeed connections is to be clear about who owns them - in the case of `accept`, the owner is the switch so the most natural approach to "stop" accepting connections is for the switch to be in control of doing so instead of having the transport pull the rug from under the switch and have the switch recover.

there's also a simple strategy to follow when doing startup and shutdown: you shut down in the reverse order that you start things - in the switch, we create the transport, then start it, then accept connections from it, so during shutdown, the natural order is to stop accepting connections, stop the transport then get rid of it.

ditto connections: it doesn't make sense to keep accepting connections _after_ stopping the transport and yet that's what the code was doing, leading to the race condition that this PR is fixing.

Reply:
Makes sense, thanks for explaining.